### PR TITLE
Move WasmState into it's own interoperation lib, so that it can be used by other filters too

### PR DESCRIPTION
**Pull request by gargnupur**
Move WasmState into WASM Api, so that it can be used by other filters too

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:Move WasmState into WASM Api, so that it can be used by other filters…
Risk Level: None
Testing: Existing tests will cover that
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: api/wasm/cpp/proxy_wasm_state.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// NOLINT(namespace-envoy)
+
+#pragma once
+
+#include <string>
+
+#include "envoy/stream_info/filter_state.h"
+
+// A simple wrapper around generic values
+class WasmState : public StreamInfo::FilterState::Object {
```
Comment (kyessenov) [inline api/wasm/cpp/proxy_wasm_state.h]: I think we need some namespace for this?

Comment (gargnupur) [reply api/wasm/cpp/proxy_wasm_state.h]: other classes in this directory are not under a namespace.. Ex: WasmData.. should we put this under namespace?

Comment (kyessenov) [reply api/wasm/cpp/proxy_wasm_state.h]: WasmState should be usable in the rest of native envoy to be able to exchange data with native extensions. I think picking some Envoy-consistent namespace is necessary for it. Other stuff here I don't have a good reason to expose.

Comment (gargnupur) [reply api/wasm/cpp/proxy_wasm_state.h]: Added Envoy::Wasm namespace... (am open to namespace name suggestions too : )

**Comment by gargnupur** [comment]
@kyessenov , @PiotrSikora, @jplevyak  : Looks like tests that failed are already failing and all the cireclci tests passed. Can you please take a look at the review?

File: api/wasm/cpp/proxy_wasm_state.h (hunk 1 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_state.h]: The API should not have any namespaces.  We are trying to standardize this and it should not be Envoy specific.

Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_state.h]: We add the namespaces when we include the file into Envoy (see wasm.cc and the way that proxy_wasm_result.h is included).

**Comment by kyessenov** [comment]
@jplevyak I wonder if it's worth having this filter state without having any common static type knowledge between the guest and the host. It's in fact just a wrapper around protobuf value. We could instead go back to dynamic metadata but not use `setDynamicMetadata` and just write the value directly into the metadata. It's effectively same as what we do with the filter state. WDYT?

**Comment by jplevyak** [comment]
RE: go back to dynamic metadata  I don't know the who history, but I think the idea is to get rid of it for other reasons (i.e. for other use cases which do have better defined data structures FilterState is going to be more efficient).

So I am inclined to keep this the way it is.

**Comment by gargnupur** [comment]
@jplevyak : Can you please help merge this? I don't have write access to merge...

**Pull request merged**